src/effects/VST3/VST3Effect.cpp

```diff
@@ -0,0 +1,787 @@
+{
+   //i18n-hint VST3 effect description string
+   return XO("SubCategories: %s").Format( mEffectClassInfo.subCategoriesString() );
+}
+
+EffectType VST3Effect::GetType()
+{
+   using namespace Steinberg::Vst::PlugType;
+   const auto& cats = mEffectClassInfo.subCategories();
+   if(std::find(cats.begin(), cats.end(), kFxGenerator) != cats.end())
+      return EffectTypeGenerate;
+   if(std::find(cats.begin(), cats.end(), kAnalyzer) != cats.end())
+      return EffectTypeAnalyze;
+   
+   return EffectTypeProcess;
+}
+
+EffectFamilySymbol VST3Effect::GetFamily()
+{
+   return VST3EffectsModule::PluginType;
+}
+
+bool VST3Effect::IsInteractive()
+{
+   return mEditController != nullptr;
+}
+
+bool VST3Effect::IsDefault()
+{
+   return false;
+}
+
+bool VST3Effect::IsLegacy()
+{
+   return false;
+}
+
+bool VST3Effect::SupportsRealtime()
+{
+   return true;
+}
+
+bool VST3Effect::SupportsAutomation()
+{
+   if(mEditController == nullptr)
+      return false;
+
+   using namespace Steinberg;
+   
+   for(int i = 0, count = mEditController->getParameterCount(); i < count; ++i)
+   {
+      Vst::ParameterInfo parameterInfo;
+      if(mEditController->getParameterInfo(i, parameterInfo) == kResultOk)
+      {
+         if(parameterInfo.flags & Vst::ParameterInfo::kCanAutomate)
+            return true;
+      }
+   }
+
+   return false;
+}
+
+bool VST3Effect::GetAutomationParameters(CommandParameters& parms)
+{
+   if(mEditController == nullptr)
+      return false;
+
+   using namespace Steinberg;
+   
+   for(int i = 0, count = mEditController->getParameterCount(); i < count; ++i)
+   {
+      Vst::ParameterInfo parameterInfo;
+      if(mEditController->getParameterInfo(i, parameterInfo) == kResultOk)
+      {
+         parms.Write(
+            std::to_string(parameterInfo.id),
+            mEditController->getParamNormalized(parameterInfo.id)
+         );
+      }
+   }
+
+   return true;
+}
+
+bool VST3Effect::SetAutomationParameters(CommandParameters& parms)
+{
+   using namespace Steinberg;
+
+   if(mComponentHandler== nullptr)
+      return false;
+
+   for(int i = 0, count = mEditController->getParameterCount(); i < count; ++i)
+   {
+      Vst::ParameterInfo parameterInfo;
+      if(mEditController->getParameterInfo(i, parameterInfo) != kResultOk)
+         continue;
+
+      Vst::ParamValue value;
+      if(parms.Read(std::to_string(parameterInfo.id), &value))
+      {
+         mComponentHandler->beginEdit(parameterInfo.id);
+         mComponentHandler->performEdit(parameterInfo.id, value);
+         mComponentHandler->endEdit(parameterInfo.id);
+         mEditController->setParamNormalized(parameterInfo.id, value);
+      }
+   }
+
+   return true;
+}
+
+bool VST3Effect::LoadUserPreset(const RegistryPath& name)
+{
+   //Dummy
+   return true;
+}
+
+bool VST3Effect::SaveUserPreset(const RegistryPath& name)
+{
+   //Dummy
+   return true;
+}
+
+RegistryPaths VST3Effect::GetFactoryPresets()
+{
+   return { };
```

Comment:
Is this also unfinished work?

I looked a little bit into VST headers and I think a "factory preset" is in their jargon a "program" (so it was in VST2) and the relevant interface for getting the information is IUnitInfo in ivstunits.h.

Reply:
It's unfinished, yes, but I don't think that "program" is a synonym for "factory preset", as it does not contain a set of parameter values, but rather enables or disables processing units